q2cli/cache.py

```diff
@@ -0,0 +1,281 @@
+    multithreaded/multiprocessing situation). Also, having a single instance
+    improves performance by only reading and/or refreshing the cache a
+    single time during its lifetime. Having two instances could, for example,
+    trigger two cache refreshes if Q2CLIDEV is set. To support these use-cases,
+    a module-level `CACHE` variable stores a single instance of this class.
+
+    """
+
+    # Public API
+
+    def __init__(self):
+        import os
+
+        # Indicates if the cache has been refreshed. For performance purposes,
+        # the cache is only refreshed a single time (at maximum) during the
+        # object's lifetime. Thus, "hot reloading" isn't supported, but this
+        # shouldn't be necessary for the CLI.
+        self._refreshed = False
+
+        self._cache_dir = self._get_cache_dir()
+
+        refresh = 'Q2CLIDEV' in os.environ
+        self._state = self._get_cached_state(refresh=refresh)
+
+    @property
+    def plugins(self):
+        """Decoded JSON object representing CLI state on a per-plugin basis."""
+        return self._state['plugins']
+
+    def refresh(self):
+        """Trigger a forced refresh of the cache.
+
+        If the cache has already been refreshed (either by this method or at
+        some point during instantiation), this method is a no-op.
+
+        """
+        if not self._refreshed:
+            self._state = self._get_cached_state(refresh=True)
+
+    # Private API
+
+    def _get_cache_dir(self):
+        import os
+        import os.path
+        import q2cli.util
+
+        cache_dir = os.path.join(q2cli.util.get_app_dir(), 'cache')
+        os.makedirs(cache_dir, exist_ok=True)
+        return cache_dir
+
+    def _get_cached_state(self, refresh):
+        import json
+        import os.path
+
+        current_requirements = self._get_current_requirements()
+        state_path = os.path.join(self._cache_dir, 'state.json')
+
+        # The cache must be refreshed in the following cases:
+
+        # 1) We have been explicitly told to refresh.
+        if refresh:
+            self._cache_current_state(current_requirements)
+        # 2) The current deployment requirements are different than the cached
+        #    requirements.
+        elif current_requirements != self._get_cached_requirements():
+            self._cache_current_state(current_requirements)
+        # 3) The cached state file does not exist.
+        elif not os.path.exists(state_path):
+            self._cache_current_state(current_requirements)
+
+        # Now that the cache is up-to-date, read it.
+        try:
+            with open(state_path, 'r') as fh:
+                return json.load(fh)
+        except json.JSONDecodeError:
+            # 4) The cached state file can't be read as JSON.
+            self._cache_current_state(current_requirements)
+            with open(state_path, 'r') as fh:
+                return json.load(fh)
```

Comment:
I'm not sure how it would fail at this point, since the cache has been recreated on line 115. It could fail if we wrote invalid JSON while regenerating the cache, but I think we can assume the cache code works as expected?